File: app/api/reviews.py
from flask import Blueprint, jsonify, request
from flask_login import login_required, current_user
from app.models import db, User, Listing, Booking, Review
from app.forms.listing_form import ListingForm, EditListingForm
from app.forms.booking_form import BookingForm
from app.forms.review_form import ReviewForm

from .auth_routes import validation_errors_to_error_messages

review_routes = Blueprint('reviews', __name__)


# GET all reviews for a listing

@review_routes.route('/listing/<int:listingId>')
@login_required
def get_reviews(listingId):
    reviews = Review.query.filter(Review.listing_id == listingId).all()
    return {'reviews': [review.to_dict() for review in reviews]}


# POST review

@review_routes.route('/listing/<int:listingId>', methods=['POST'])
@login_required
def add_review(listingId):
    form = ReviewForm()
    form['csrf_token'].data = request.cookies['csrf_token']
    if form.validate_on_submit():
        review = Review(
            listing_id = listingId,
            guest_id = current_user.id,
            rating = form.rating.data,
            comment = form.comment.data
        )
        db.session.add(review)
        db.session.commit()
        return review.to_dict()
    return {'errors': validation_errors_to_error_messages(form.errors)}, 401


# EDIT review

@review_routes.route('/<int:id>', methods=['PUT'])
@login_required
def edit_review(id):
    form = ReviewForm()
    review = Review.query.get(id)
    form['csrf_token'].data = request.cookies['csrf_token']
    if form.validate_on_submit():
        review.rating = form.rating.data
        review.comment = form.comment.data
        db.session.commit()
        return review.to_dict()
    return {'errors': validation_errors_to_error_messages(form.errors)}, 401


# DELETE review

@review_routes.route('/<int:id>', methods=['DELETE'])
@login_required
def delete_review(id):
    review = Review.query.filter_by(id=id).first()
    db.session.delete(review)
    db.session.commit()
    return review.to_dict()

File: react-app/src/store/reviews.js
const LOAD_REVIEW = 'listing/LOAD_REVIEW'
const ADD_REVIEW = 'listing/ADD_REVIEW'
const EDIT_REVIEW = 'listing/EDIT_REVIEW'
const REMOVE_REVIEW = 'listing/REMOVE_REVIEW'

export const loadReviews = reviews => ({
    type: LOAD_REVIEW,
    reviews
})

export const postReview = review => ({
    type: ADD_REVIEW,
    review
})

export const editReview = (review, reviewId) => ({
    type: EDIT_REVIEW,
    review,
    reviewId
})

export const deleteReview = (review) => ({
    type: REMOVE_REVIEW,
    review
})

export const getReviews = listingId => async dispatch => {
    const res = await fetch(`/api/reviews/listing/${listingId}`)

    if (res.ok) {
        const reviews = await res.json()
        dispatch(loadReviews(reviews))
    }
}

export const addReview = (listingId, rating, comment) => async dispatch => {
    const res = await fetch(`/api/reviews/listing/${listingId}`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            rating,
            comment
        })
    })

    if (res.ok) {
        const data = await res.json()
        dispatch(postReview(data))
    }
}

export const updateReview = (reviewId, rating, comment) => async dispatch => {
    const res = await fetch(`/api/reviews/${reviewId}`, {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            rating,
            comment
        })
    })

    if (res.ok) {
        const review = await res.json()
        dispatch(editReview(review, reviewId))
    }
}

export const removeReview = (reviewId) => async dispatch => {
    const res = await fetch(`/api/reviews/${reviewId}`, {
        method: 'DELETE'
    })

    if (res.ok) {
        const review = await res.json()
        dispatch(deleteReview(review))
    }
}

export default function reviewReducer(state = [], action) {
    switch(action.type) {
        case LOAD_REVIEW:
            return action.reviews
        case ADD_REVIEW:
            return [ ...state, action.review ]
        case EDIT_REVIEW:
            return state.map(review => {
                if (review.id === action.reviewId) {
                    return action.review
                }
                return review
            })
        case REMOVE_REVIEW:
            return state.filter(review => review.id !== action.review.id)
        default:
            return state
    }
}
